doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,314 @@
+   `Coordinator` thread.
+2. Allows the `StorageController` to pick the timestamp for the update, instead of the
+   `Coordinator`, which aligns with our long term goals if having a storage owned "push source" in
+   the future.
+
+### Request Validation
+[request_validation]: #request_validation
+
+Every `WEBHOOK` source will be open to the entire internet, as such we need to include some sort of
+validation so we can verify that requests are legitimate. In general webhooks are validated by
+HMAC-ing the request body using SHA256, and validating the result with a signature provided in the
+request headers. An issue though is everyone does this just a little bit different:
+
+* Segment signature = HMAC of just the body of the request
+* Buildkite signature = HMAC of "#{timestamp}.#{body}"
+* GitHub signature = "body=" + HMAC of {body}
+
+As such we'll need to support some custom logic for validation. What we can do is support a
+`VALIDATE USING` statement that accepts only a single scalar expression. This expression will
+be provided the request `headers` as `map[text] -> text`, and the `body` as `text` _regardless of
+whether or not the source has `INCLUDE HEADERS` or what `FORMAT BODY` is specified._ For example:
+
+```
+VALIDATE USING (
+  headers['X-Signature'] = hmac('sha256', SECRET webhook_secret, body)
+)
+```
+
+Internally the provided expression will be turned into a `MirScalarExpr` which will be used to
+evaluate each request. This evaluation will happen off the main coordinator thread.
+
+### Request Limits
+[request_limits]: #request_limits
+
+To start we'll aim to support at least 100 requests per second, in other words, 10 sources could
+receive 10 requests per second. We can probably scale further, but this seems like a sensible limit
+to start at. Also we'll add an [`axum::middleware`](https://docs.rs/axum/latest/axum/middleware/index.html)
+that enforces rate limiting on our new endpoint. This should be a pretty easy way to prevent a
+misbehaving source from taking down all of `environmentd`.
+
+We'll also introduce a maximum size the body of each request can be, the default limit will be
+16KiB.
+
+Both the maximum number of requests per second, and max body size, will be configurable via
+LaunchDarkly, and later we can introduce SQL syntax so the user can `ALTER` these parameters on a
+per-source basis.
+
+> **_Note_**: Some applications provide batching of webhook events. To reduce scope we are not
+going to support this, our current thought is handling batched requests should come naturally with
+support for `TRANSFORM USING`.
+
+### Metrics
+[metrics]: #metrics
+
+There are three metrics we should track, to measure success and performance of this new source:
+
+1. Number of `WEBHOOK` sources created.
+2. Number of requests made to the `/api/webhook` endpoint.
+    * Subdivided by number of successes and failures.
+3. Response time of the `/api/webhook` endpoint.
+
```

Comment:
I think it'd be great to quickly describe the implementation of these metrics.

```suggestion
Metric 1 will be collected via the Prometheus SQL exporter running a query like `SELECT count(*) FROM mz_sources WHERE type = 'webhook';`

Metric 2 will be a new Prometheus counter called `mz_webhook_requests`.

Metric 3 will be a new Prometheus histogram named `mz_webhook_request_duration_seconds`.
```

Feel free to workshop the names.

Reply:
Good idea, updated!